### ISSUE TITLE
Synchronization between SML and Soar input link off?

### ISSUE THREAD

**Issue by moschmdt** [SoarGroup/Soar#418]
I am currently working on integrating ROS2 service and clients in Soar and observed some interesting behavior:

1. Attach new information on Soar input-link via SML, e.g. `ObjectDetected`
2. Soar Debugger shows input, is processed, and `^status complete` is added in Soar to `ObjectDetected`.
3. `^status complete` is also visible in the Soar debugger.
4. Iterating over all children of the pointer to that input element in SML does not show the `status complete` child WME. Expected behavior: `status complete` WME is also listed in SML.

I ensured the reference in SML to the WME is correct and also called `pAgent->runSelf(1)` repeatedly, and implemented the proposed environment model with `pAgent->RunAllAgentsForever()` in a separate thread. 

Example output of *SML*:
```bash
[gdbserver-1] [INFO] [1701167445.264280157] [soar_agent]: (-6: g2 ^ObjectDetected o5)children:
[gdbserver-1]   (-7: o5 ^name basin)
[gdbserver-1]   (-8: o5 ^material mineralcast)
[gdbserver-1]   (-9: o5 ^workspace_name default)
[gdbserver-1]   (-10: o5 ^id -1625742390542062751)
```

*Soar Debugger*:
```bash
$ print -i --depth 10 I1

(12: I1 ^input-link I2)
(13: I1 ^output-link I3)
  (14: I2 ^gui G1)
  (15: I2 ^robot R4)
  (16: I2 ^visual V1)
    (97: G1 ^ObjectDetected O4)
      (101: O4 ^id -1625742390542062751)
      (99: O4 ^material mineralcast)
      (98: O4 ^name basin)
      (118: O4 ^status complete)
      (100: O4 ^workspace_name default)
```

Is there any synchronization that must be called explicitly between SML, client, and kernel?

**Comment by garfieldnate** [comment]
Just writing to tell you that I saw this but I don't know precisely what causes the issue. There are no explicit synchronization functions in Soar as far as I know. It's a little unclear to me exactly how you are hitting this issue. Which thread is running what, and which one can't see the `^status complete` WME?

Also, I see (`AddStatusComplete`)[https://github.com/search?q=repo%3ASoarGroup%2FSoar%20AddStatusComplete&type=code] in SML, but in Soar it only appears to be called in tests, which implies to me that it's the client's job to call it. Have I misunderstood that, or are you calling it on a thread somewhere?

**Comment by moschmdt** [comment]
Well, it is unclear to me as well...

1. SML thread (communication to ROS/ environment): Attaches the object to the input link. 
1. Soar Kernel (via `CreateKernelInNewThread`) thread: Attaches `^status ...` to the WME on the input-link based on some rules. So the `AddStatusComplete` is never called from the client (SML), instead the status complete is added from Soar to the input-link WME.

I confirmed via the Debugger that the status complete is added on the input link, but in SML (Cpp) I can't see it. So essentially, I can "write" to the input-link, but can´t read from it.

I hope that clarifies the different threads?

**Comment by garfieldnate** [comment]
The SML quick-start guide shows the ^status complete stuff being used for commands sent *from* Soar via the output link, not *to* Soar via the input link.

I think it's likely that a Soar agent can't make changes to the input link to send information to the outside world, as this is what the output link is for. So I guess if you're sending commands to Soar via the input link, and you want to see which ones have completed, you could add a rule to your agent that notifies the outside world of the completion via the output link, and I'm certain that would work.

Similarly, I would expect it to not work if you tried to send information to the Soar agent by editing the output link.

One on our team likes to say that Soar wants to "own the world," and I think this is an effect of that. We expect Soar to be issuing the commands to the outside environment, which then reflects the state of the world on the input link.

I'll try to confirm with someone on this, but I think we should change this to a "documentation" ticket, because this is likely expected behavior.

**Comment by garfieldnate** [comment]
This makes more sense the more I think about it. SML is a communication layer between Soar and other software. There's an interface used within Soar (kernel SML) to interact via SML with the outside world, and there's an interface exposed for clients (client SML) for interacting with Soar. The client SML part sends information added to the input link. The kernel SML part sends information added to the output link. Each side owns the link that they use to send data, and it would make sense that the opposite side cannot change this link.

See Bob's comment [here](https://github.com/SoarGroup/Soar/issues/375#issuecomment-1696476091):

>As a rule of thumb, soar owns main working memory and the output link, the application owns the input link.

**Comment by moschmdt** [comment]
Thank you for the explanation! I agree with you that this design decision makes sense, but then I would expect some kind of warning/ error etc:

- as soon as a production is loaded that is supposed to attach anything (both O/I-supported) to the input link, the agent should crash or at least throw an error
- The debugger should NOT show any WME generated by Soar on the input ink.

**Comment by marinier** [comment]
There's nothing wrong with rules putting WMEs on the input link. They don't
get communicated outside the agent, but we do this sort of thing all the
time to, e.g., mark an input message as processed so the operator that does
the processing can unmatch.

**Comment by garfieldnate** [comment]
Given Bob's comment, I'm not sure that there's much else we can do besides improve documentation. Added warnings would be spurious for many users, and not showing the WME in the debugger would also be undesirable behavior when designing agents that depend on changing the input link. I always prefer to design things in a way that the documentation is superfluous, though, so other ideas are welcome. I thought about maybe a flag or font change or something in the debugger, but this would probably be a very large effort for what I believe is an uncommon use case.

When writing notes about a technology, I do like to have something like a "gotcha" or "footgun" document that lists common pitfalls and known issues. If we had a footgun document for Soar, I would definitely put this (and the linked issue) in there.

**Comment by moschmdt** [comment]
Regarding the documentation: I found something on the website while converting it to Markdown: 
- [Design Dogma](https://soar.eecs.umich.edu/articles/articles/faqs-and-guides/77-soar-design-dogma).
- [FAQ](https://soar.eecs.umich.edu/articles/articles/faqs-and-guides/75-soar-technical-faq) even states that the input link can be manipulated (what exactly manipulation means w.r.t. this issue is unclear in my eyes and weather this includes reading capabilities). The documentation is almost 10 years old - Is it still applicable?

**Comment by marinier** [comment]
The design dogma is closer to 20 years old. It was written by Andy Nuxoll
when we were in grad school and it was his opinion at the time. This was
before all the newer memories and learning mechanisms. It's probably not
wrong, but shouldn't be taken as endorsed by John or anything.

The input link can be manipulated by both the IO system and rules. It's
just that the changes made by rules can't be seen by the IO system.

Sort of similar to the dogma and much newer (but still quite old) is the
soar programmer's guide. I wrote this to capture some patterns I've used
repeatedly, but haven't revisited it in forever. I think John did not agree
with everything I said in there.
https://soartech.github.io/SoarProgrammersGuide/31654126.html
